Interned strings unification for TS/NTS

weltling opened the pull request:
The primary goal is to support real interned strings also in Ts builds. The unified approach implements string interning with two tables for permanent and per-request interned strings for both TS/NTS. More explanations to the patch here http://news.php.net/php.internals/98324

Thanks.

dstogov commented on main/streams/streams.c:
```diff
@@ -1706,7 +1706,7 @@ PHPAPI int php_register_url_stream_wrapper(const char *protocol, php_stream_wrap
 	}
 
-	return zend_hash_str_add_ptr(&url_stream_wrappers_hash, protocol, protocol_len, wrapper) ? SUCCESS : FAILURE;
+	return zend_hash_add_ptr(&url_stream_wrappers_hash, zend_string_init_interned(protocol, protocol_len, 1), wrapper) ? SUCCESS : FAILURE;
```
This place may be reached at run-time through stream_wrapper_register(), but we don't create interned strings at run-time.

weltling replied:
I was basing it on the fact url_stream_wrappers_hash is used, which is global. Otheriwse, stream wrappers created on runtime seem to be not thread safe in general. Gonna investigate and rework this.

weltling replied:
Seems this part is ok. User space wrappers runs through php_register_url_stream_wrapper_volatile() and save into FG(stream_wrappers). So far i couldn't find a code path where php_register_url_stream_wrapper() is called outside the startup phase. Thus names of those saved into url_stream_wrappers_hash need to be interned, as they can interact with the runtime code.

laruence commented on Zend/zend_string.c:
```diff
@@ -78,57 +88,85 @@ void zend_known_interned_strings_init(zend_string ***strings, uint32_t *count)
-#endif
+	empty_string = zend_new_interned_string_permanent_int(str);
+#ifndef ZTS
+	CG(empty_string) = empty_string;
```
since empty_string is a true global variable,  seems we don't need to store it in CG .

laruence commented on TSRM/TSRM.c:
```diff
@@ -170,6 +177,11 @@ TSRM_API void tsrm_shutdown(void)
+		return;
+	}
+
 	if (tsrm_tls_table) {
```
as we already have a in_main_thread guarder, this check should also not that necessary, right?

weltling replied:
I'll check this, thanks for the hint. Thought also to export this as a function, so it's usable in any places with startup/shutdown, maybe improving to explicitly checking thread ids.

weltling replied:
This seems ok, i've reverted for now as it's not related to interned strings directly. Will re-apply once the actual interned strings stuff is finished.

dstogov commented on TSRM/TSRM.c:
```diff
@@ -10,7 +10,7 @@
 */
 
-#include "TSRM.h"
+#include "zend.h"
```
TSRM didn't depend on Zend by design. Can we avoid this?

weltling replied:
We could introduce more callbacks in TSRM, currently there are only a couple of callbacks at thread creation. Some callbacks could be introduced for the tsrm startup/shutdown and thread shutdown. I'll do the callbacks for now, for cleaner TSRM API.

dstogov commented on TSRM/TSRM.c:
```diff
@@ -170,30 +177,33 @@ TSRM_API void tsrm_shutdown(void)
+	if (!in_main_thread) {
+		/* ensure singleton */
+		return;
+	}
```
Changes/fix related to "tsrm_tls_table" and "in_main_thread" look unrelated to interned strings. Lets commit this part separately.

weltling replied:
ack, will revert for now.

dstogov commented on Zend/zend.c:
```diff
@@ -623,7 +619,7 @@ static void executor_globals_ctor(zend_executor_globals *executor_globals) /* {{
 
 	zend_startup_constants();
-	zend_copy_constants(EG(zend_constants), GLOBAL_CONSTANTS_TABLE);
+	zend_copy_constants(executor_globals->zend_constants, GLOBAL_CONSTANTS_TABLE);
```
This change is unrelated. Commit it separately.

dstogov commented on Zend/zend.c:
```diff
@@ -815,17 +817,16 @@ int zend_startup(zend_utility_functions *utility_functions, char **extensions) /
 
 	zend_hash_destroy(executor_globals->zend_constants);
-	*executor_globals->zend_constants = *GLOBAL_CONSTANTS_TABLE;
+	executor_globals->zend_constants = GLOBAL_CONSTANTS_TABLE;
```
I'm not sure about all these zend_startup() changes above.

dstogov commented:
I see a serious problem with your patch.

The concept of "interned strings" assumes a single copy of each particular string. This allows checks interning string for equality just comparing the pointers instead of full memcmp(). (see zend_hash_find_bucket() for example).

Your patch introduces two (independent) interned strings tables and as result we may create two copies of each string, wasting memory and time doing memcmp().

This may be fixed, if we make hierarchy of interned string tables.
1) Permanent
2) Compile-time
If we add a new interned string into "compile-time" table, we first should check if it's already exists in "Permanent" and only then add into "Compile-Time" table.

This design decision may also simplify opcache support for interned string (instead of copying all permanent interned strings into SHM we may add another interned string table level between (1) and (2)).

nikic commented on Zend/zend_string.c:
```diff
@@ -78,71 +88,93 @@ void zend_known_interned_strings_init(zend_string ***strings, uint32_t *count)
-		return str;
+	/* Handle the passed string as read only. */
+	h = ZSTR_H(str);
+	if (!h) {
```
Doesn't this imply that strings won't get deduplicated if they don't happen to have their hash computed by some unrelated previous lookup?

weltling replied:
Indeed. I was worried about a possible race condition, but zend_string_forget_hash_val() is currently used in thread safe manner, and otherwise an interned string will always have a hash val. Fixing.

weltling commented:
I think it's now ready for the next review round. Things done

- reverted startup/shutdown related changes in zend.c
- implemented look up with the table hierarchy, as per Dmitry's suggestion
- implemented SHM table within the hierarchy, as per Dmitry's suggestion

The pointers to the strings from the persistent table are not replaced by Opcache anymore. With this, no dangling pointers are anymore in the main thread after the Opcache detached the memory. This is cleaner in any case, even if we repeat the persistent table per process. My extensive tests on Linux/Windows with thread safe Apache MPMs are green so far as well.

Thanks.

nikic commented on Zend/zend_API.c:
```diff
@@ -2019,7 +2019,7 @@ ZEND_API zend_module_entry* zend_register_module_ex(zend_module_entry *module) /
 	zend_str_tolower_copy(ZSTR_VAL(lcname), module->name, name_len);
 
-	if ((module_ptr = zend_hash_add_mem(&module_registry, lcname, module, sizeof(zend_module_entry))) == NULL) {
+	if ((module_ptr = zend_hash_add_mem(&module_registry, zend_new_interned_string(lcname), module, sizeof(zend_module_entry))) == NULL) {
```
This should be `lcname = zend_new_interned_string(lcname)` on a separate line. The semantics of zend_new_interned_string are similar to realloc. The argument is released, you just sometimes don't see it (if the string was not interned yet).

nikic commented on Zend/zend_builtin_functions.c:
```diff
@@ -763,7 +763,7 @@ ZEND_FUNCTION(each)
 	}
 	zend_hash_index_add_new(Z_ARRVAL_P(return_value), 1, entry);
-	zend_hash_add_new(Z_ARRVAL_P(return_value), CG(known_strings)[ZEND_STR_VALUE], entry);
+	zend_hash_add_new(Z_ARRVAL_P(return_value), ZSTR_KNOWN(ZEND_STR_VALUE), entry);
```
Is it possible to apply these changes changes independently? (Only if it's not too entangled.)

weltling replied:
CG(known_strings) is now removed, instead it's a real global var zend_known_strings. Same is with zend_empty_string and zend_one_char_string. So we would reference a global var directly, which is worse than having an acessor macro. But in general, it is a a change that belongs to the patch, so either way it were not possible to keep the exact current code.

nikic commented on Zend/zend_string.c:
```diff
@@ -72,175 +57,207 @@ ZEND_KNOWN_STRINGS(_ZEND_STR_DSC)
+	interned_strings->nTableMask = -interned_strings->nTableSize;
+	HT_SET_DATA_ADDR(interned_strings, pemalloc(HT_SIZE(interned_strings), permanent));
+	HT_HASH_RESET(interned_strings);
+	interned_strings->u.flags |= HASH_FLAG_INITIALIZED;
```
Isn't this code just `zend_hash_real_init(interned_strings, 0);`?

nikic commented on Zend/zend_string.c:
```diff
@@ -72,175 +57,207 @@ ZEND_KNOWN_STRINGS(_ZEND_STR_DSC)
+			memcpy(interned_strings->arData, old_buckets, sizeof(Bucket) * interned_strings->nNumUsed);
+			pefree(old_data, interned_strings->u.flags & HASH_FLAG_PERSISTENT);
+			zend_hash_rehash(interned_strings);
+		} else {
```
With `pemalloc` this branch can never be taken.

nikic commented on Zend/zend_string.c:
```diff
@@ -72,175 +57,207 @@ ZEND_KNOWN_STRINGS(_ZEND_STR_DSC)
+
+static zend_never_inline void zend_string_table_grow(HashTable *interned_strings)
+{
+	if (EXPECTED(interned_strings->nTableSize < HT_MAX_SIZE)) {	/* Let's double the table size */
```
I know this is existing code, but the implementation as-is cannot actually handle this condition gracefully. Either it should error as the HT would usually do, or it should not intern in that case.

weltling replied:
Probably it should fail hard in this situation, as not interning might cause more hard weight issues, at least with TS. Alternatively, we could increase the value, if possible.

nikic commented on Zend/zend_string.c:
```diff
@@ -72,175 +57,207 @@ ZEND_KNOWN_STRINGS(_ZEND_STR_DSC)
-				CG(interned_strings).nTableSize = CG(interned_strings).nTableSize >> 1;
-				CG(interned_strings).nTableMask = -CG(interned_strings).nTableSize;
-			}
+	return NULL;
```
As the interned string is stored as both key and value of the HT (https://github.com/weltling/php-src/blob/97882476c602a20411c33d0d5c555818ba5e5834/Zend/zend_string.c#L176) this function should be a thin wrapper around `zend_hash_find()`, instead of reimplementing HT logic.

nikic commented on Zend/zend_string.c:
```diff
@@ -72,175 +57,207 @@ ZEND_KNOWN_STRINGS(_ZEND_STR_DSC)
-	HT_HASH(&CG(interned_strings), nIndex) = HT_IDX_TO_HASH(idx);
+	nIndex = h | interned_strings->nTableMask;
+	Z_NEXT(p->val) = HT_HASH(interned_strings, nIndex);
+	HT_HASH(interned_strings, nIndex) = HT_IDX_TO_HASH(idx);
```
This could use `zend_hash_append()`. Or going one step further, this could just use `zend_hash_add_new()`, in which case we can discard the duplicated table growing logic as well.

weltling replied:
It looks like the only not addressed points regard to the interned strings ht management. I would address them after the patch was merged. It would be indeed make sense to reuse the existing API, for now we're fine reusing the existing implementation.

nikic replied:
Sounds good. Patch LGTM, very nice work!

nikic commented on ext/opcache/ZendAccelerator.c:
```diff
@@ -505,39 +489,35 @@ zend_string *accel_new_interned_string(zend_string *str)
-#ifndef ZTS
 /* Copy PHP interned strings from PHP process memory into the shared memory */
-static void accel_use_shm_interned_strings(void)
+static void accel_copy_permanent_strings(zend_new_interned_string_func_t new_interned_string)
```
Just to clarify: Copying to SHM here is just an optimization right? For cross-process purposes it's only important that strings referenced from SHM are interned, not that permanent strings are all in SHM. If something is missed here, it should not cause issues. Is that correct?

weltling replied:
On ASLR enabled systems, like Windows by default or hardened Linux, pointers to interned strings are likely differ. So in first place, by copying we ensure the pointers are same in all the processes. If something is missed here, the interning will happen on demand in zend_persist.c and zend_persist_calc.c parts on demand. See ADD_INTERNED_STRING, zend_accel_store_interned_string and so on there.

It might be, that on some systems the pointers to the local interned strings could stay same, But as we can't guarantee it, it's played safe. While it were possible to add a configure option and modify the code for not saving to SHM, i'd not go for that, as it is really not a reliable behavior.

nikic commented on ext/opcache/ZendAccelerator.c:
```diff
@@ -582,40 +562,46 @@ static void accel_use_shm_interned_strings(void)
+			p->key = new_interned_string(p->key);
+		}
+		c = (zend_constant*)Z_PTR(p->val);
+		if (c->name) {
```
Are there constants without names?

weltling replied:
Probably not a usual case. Some tricks i recall were like giving a name with a \0-byte in the front, to save some thing unusual into the constant. Such code is for sure still somewhere in APCu and earlier in APC, probably in some other exts as wel.

nikic commented on ext/opcache/ZendAccelerator.c:
```diff
@@ -629,13 +615,70 @@ static void accel_use_shm_interned_strings(void)
+		}
+	}
+
+	for (idx = 0; idx < module_registry.nNumUsed; idx++) {
```
All of this code should be using ZEND_HASH_FOREACH_BUCKET... but not related to this change.

dstogov commented:
We have to copy permanent interned strings into SHM, because on Windows different processes may put the same strings at different addresses.

________________________________
From: Nikita Popov <notifications@github.com>
Sent: Friday, March 3, 2017 12:45:54 AM
To: php/php-src
Cc: Dmitry Stogov; Comment
Subject: Re: [php/php-src] Interned strings unification for TS/NTS (#2390)


@nikic commented on this pull request.

________________________________

In ext/opcache/ZendAccelerator.c<https://github.com/php/php-src/pull/2390#discussion_r104037654>:

>  /* Copy PHP interned strings from PHP process memory into the shared memory */
-static void accel_use_shm_interned_strings(void)
+static void accel_copy_permanent_strings(zend_new_interned_string_func_t new_interned_string)


Just to clarify: Copying to SHM here is just an optimization right? For cross-process purposes it's only important that strings referenced from SHM are interned, not that permanent strings are all in SHM. If something is missed here, it should not cause issues. Is that correct?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/php/php-src/pull/2390#pullrequestreview-24851710>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACZM0nEuHBvRQBpb_VfqTT7WZ_k2Su5Oks5rhziSgaJpZM4MEM_p>.

weltling commented:
Merged per c6982995504b6e21e8a5ade29cfb16a55196dc43. The fololwups are going to be handled either directly in dev or per a separate PR.

Thanks!
